### PR TITLE
System csparse

### PR DESCRIPTION
**Pull request by radarsat1**
Towards #189.

Set USE_SYSTEM_SUITESPARSE=ON to use `cs.h` in `/usr/include/suitesparse` from the `libsuitesparse-dev` package in Debian/Ubuntu.

Important caveat: the modified CSparse in Siconos redefines the integer types depending on `SICONOS_INT64`.  In `cs.h` from libsuitesparse-dev, the structs use `int`.  So this patch contains,

    typedef int csi;

in `CSparseMatrix.h`, since `csi` is used in many places in Numerics.  I point this out because I am not sure of the full implications of no longer supporting `int64_t`.  Apart from that it redefines CSparseMatrix as,

    typedef struct cs_di_sparse CSparseMatrix

instead of `struct cs_sparse`.

Does not seem to change test results but still thought I should propose as a PR for comments due to the above uncertainties about the integer type.  Tested with and without BUILD_AS_CPP, some bonus patches in this PR contain a couple of fixes for that.

The mechanism is that when you se USE_SYSTEM_SUITESPARSE, it looks for cs.h, and sets the flag SICONOS_USE_CXSPARSE.  CXSparse is an extended version of CSparse that is distributed by the libsuitesparse-dev package.  Apart from renaming the `struct cs_di_sparse`, this actually was lot of less work than I expected, so probably something is broken ;)

### CODE REVIEW

**Comment by vacary** [comment]
Perfect.

I am wondering if it could be useful to set csi directly to use CSXSPARSE https://github.com/jluttine/suitesparse/tree/master/CXSparse

Just an open question.

**Comment by radarsat1** [comment]
Aaah, thank you for pointing that out I missed it completely!  Indeed there are some things there that maybe we can use.

File: externals/CMakeLists.txt [hunk 1]
```diff
@@ -86,7 +110,7 @@ set(${COMPONENT}_LINKER_LANGUAGE C)
 if(MLCPSIMPLEX_FOUND) 
   set(${COMPONENT}_LINKER_LANGUAGE CXX)
 endif()
-set(${COMPONENT}_LINK_LIBRARIES ${SICONOS_LINK_LIBRARIES})
+set(${COMPONENT}_LINK_LIBRARIES ${SICONOS_LINK_LIBRARIES} ${CXSparse_LIBRARY})
```
Comment (xhub) [inline externals/CMakeLists.txt]: This doesn't look right. The library should be added to SICONOS_LINK_LIBRARIES. I believe this happens in the compile_with macro.

**Comment by radarsat1** [comment]
I've got this update ready, but I'm a bit preoccupied because I noticed that valgrind is reporting a memory leak in test_sparse that it wasn't before.  I think it has something to do with the cs_sparse struct being renamed, but I haven't figured it out yet.  Maybe I'll push the to this branch in the meantime.

**Comment by radarsat1** [comment]
Pushed, but I suspect there may still be an issue in `numerics_NM.i`

File: externals/CMakeLists.txt (hunk 1 above)
Comment (radarsat1) [reply externals/CMakeLists.txt]: Yeah, please review this heavily ;)  I'll take a look.

**Comment by radarsat1** [comment]
Oh okay, not so straightforward.  On my Ubuntu laptop (17.04), I get:

    $ valgrind python -c 'import test_sparse as t; t.test_SBM_from_sparse1()'

~~~
==17337== LEAK SUMMARY:
==17337==    definitely lost: 0 bytes in 0 blocks
==17337==    indirectly lost: 0 bytes in 0 blocks
~~~

with both `USE_SYSTEM_SUITESPARSE=YES` and `NO`.

Seems to be okay.  I saw a problem previously under Docker (Ubuntu 14.04)

~~~
==7621== LEAK SUMMARY:
==7621==    definitely lost: 80 bytes in 1 blocks
==7621==    indirectly lost: 492 bytes in 2 blocks
~~~

but now after pushing I restarted the container from scratch and am getting 0 bytes.  So maybe it is fixed.  Hate it when that happens.  I'll keep investigating.

File: numerics/swig/Numerics_typemaps_numericsmatrices.i [hunk 2]
```diff
@@ -262,10 +262,18 @@
...
-#define CSparseMatrix cs_sparse
+#ifdef SICONOS_USE_CXSPARSE
+#define CSparseMatrix cs_di_sparse
+#else
+#define CSparseMatrix cs_sparse;
+#endif
+#else
+#ifdef SICONOS_USE_CXSPARSE
+#define CSparseMatrix struct cs_di_sparse
 #else
```
[the remark this answers is not in the document]
Comment (radarsat1) [reply numerics/swig/Numerics_typemaps_numericsmatrices.i]: Yeah the CS_LONG stuff takes a little thought.  Actually we _do_ have control over it as far as I can tell.  That is, CXSparse exports both int and long versions (cs_di_sparse, cs_dl_sparse), and based on the value of CS_LONG that is set when Siconos is compiled, the cs_* functions will expand to the right version.  So SICONOS_INT64 seems to be the right way to control which CS_LONG we wish to use.  To interact with Siconos, the user must ensure he uses the same value of CS_LONG as was used when Siconos was compiled.  At least I think this is the case.  Hence I considered it okay to set in SparseMatrix.h based on the value of SICONOS_INT64 --- users of SparseMatrix.h _must_ use the value of CS_LONG that was specified during compilation, i.e., the value of SICONOS_INT64 in SiconosConfig.h.

Unless of course we want to make "di" and "dl" versions of all functions that take CSparseMatrix.

> Using CS_NAME(sparse) instead of hardcoding CSparseMatrix to a specific integer type would be nice.

I did try this actually, I was using a typedef to `CNAME(_sparse)` before, but swig didn't like it, gave syntax errors.

**Comment by radarsat1** [comment]
> I think all the CSparse stuff should be in-as-much hidden from the user. Ideally it would have to include a header file himself, for advance usage.

Yes, I proposed this in #188.  I changed the `#include "csparse.h"` in SparseMatrix.h to an opaque `struct cs_sparse`, and the user would have to `#include` it himself to access the struct.  We could still do this with CXSparse perhaps.  Mainly I thought it was a little unusual to include `csparse.h` in the Siconos installed files and for Siconos to export the `cs_` symbols.

That said, I am not sure I understand your concern about using SuiteSparse/CXSparse as a library.  Or should I understand that CXSparse is better to use as a library than CSparse?  Fwiw, I am proposing to use the .so files found in libsuitesparse-dev, which only includes CXSparse, not CSparse.  (I am not sure why this package is not split into binary and header-file packages but whatever.)

I don't know much about compiling .mex, does Matlab include its own version of  `cs.h`? I assume by "clash" you mean that both Siconos and Matlab were _exporting_ the cs symbols. If Siconos is now _importing_, I don't see a problem.
Moreover ,if Matlab re-exports the `cs_` functions perhaps it's possible to link to those instead.  

By the way speaking of symbol exports, there seems to be a trend in Debian to control symbol exports more carefully, we may have to deal with that.  [(Gcc  "visibility" attribute.)](https://gcc.gnu.org/wiki/Visibility)  It might be a bit of a headache eventually.

File: numerics/src/FrictionContact/Friction_cst.h [hunk 3]
```diff
@@ -161,8 +161,6 @@ extern const char* const   SICONOS_GLOBAL_FRICTION_3D_GAMS_PATHVI_STR;
 extern const char* const   SICONOS_GLOBAL_FRICTION_3D_VI_FPP_STR;
 extern const char* const   SICONOS_GLOBAL_FRICTION_3D_VI_EG_STR;
 extern const char* const   SICONOS_FRICTION_3D_ONECONTACT_QUARTIC_STR ;
-extern const char* const   SICONOS_GLOBAL_FRICTION_3D_GAMS_PATHVI_STR;
-extern const char* const   SICONOS_FRICTION_3D_ONECONTACT_QUARTIC_STR ;
 extern const char* const   SICONOS_FRICTION_3D_ONECONTACT_QUARTIC_NU_STR ;
```
[the remark this answers is not in the document]
Comment (radarsat1) [reply numerics/src/FrictionContact/Friction_cst.h]: > [numerics] fix "Redundant redeclaration" warnings.

They were defined twice.

File: cmake/siconos_find_common_dep.cmake [hunk 4]
```diff
@@ -267,12 +267,6 @@ if(UNIX)
 endif()
 
 
-# SiconosConfig.h generation and include
-if(EXISTS ${CMAKE_SOURCE_DIR}/config.h.cmake)
-  configure_file(${CMAKE_SOURCE_DIR}/config.h.cmake
-    ${CMAKE_BINARY_DIR}/SiconosConfig.h)
-endif()
-
 set(${PROJECT_NAME}_LOCAL_INCLUDE_DIRECTORIES
```
[the remark this answers is not in the document]
Comment (radarsat1) [reply cmake/siconos_find_common_dep.cmake]: I moved the config generation to the end of CMake instead of the beginning, otherwise I could not add `SICONOS_USE_CXSPARSE`, which I define during `externals/CMakeLists.txt`.  It is more usual to generate config as the last step in a build system, no?  In any case we can debate but it wasn't obvious to me that config generation happens during `siconos_find_common_dep`.  Otherwise, SuiteSparse/CXSparse detection could happen in that file instead, but it seemed to make sense to me to do it in externals.

File: numerics/src/LCP/LinearComplementarityProblem_as_ConvexQP.c [hunk 5]
```diff
@@ -18,7 +18,7 @@
 */
 #include <stdlib.h>
 #include <assert.h>
-#include <math.h>
+#include <tgmath.h>
 #include "LinearComplementarityProblem_as_ConvexQP.h"
```
[the remark this answers is not in the document]
Comment (radarsat1) [reply numerics/src/LCP/LinearComplementarityProblem_as_ConvexQP.c]: Perhaps not.  Previously these files used `max` not `fmax` so I was trying to maintain semantics since `max` doesn't compile for me with BUILD_AS_CPP.

> /home/sinclairs/projects/siconos/numerics/src/LCP/lcp_compute_error.c:52:45: error: no matching function for call to ‘max(int, double)’

However yes I can confirm that changing `max` to `fmax` and not changing `math.h` to `tgmath.h` also works.

File: externals/CMakeLists.txt (hunk 1 above)
Comment (radarsat1) [reply externals/CMakeLists.txt]: I'll move SuiteSparse configuration stuff to `siconos_find_common_dep`.

**Comment by radarsat1** [comment]
Okay, I think I fixed most of the concerns.  Continuing to test locally, especially regarding SWIG stuff.

The main concern is that we have this `%extend` directive for allocating and freeing `struct cs_sparse` properly, that place in `numerics_NM.i` where @xhub [wrote](https://github.com/siconos/siconos/blob/d5771e0e5f6c5516309f2dea48069f15ef8265d0/numerics/swig/numerics_NM.i#L138),

> I'm not sure it's a good idea to duplicate thise here ... it is already defined in csparse.h

following a definition of `struct cs_sparse`.  I assume this was to avoid doing `#include "csparse.h"` and therefore exporting all of CSparse in numerics.py.  However, with CXSparse, that struct doesn't exist.

I [added](https://github.com/siconos/siconos/pull/190/files#diff-40f3a7487a1955a014a659a91e830857L138), a bit awkwardly,

```C++
#ifdef SICONOS_USE_CXSPARSE
#ifdef SICONOS_INT64
#define cs_sparse cs_dl_sparse
#else
#define cs_sparse cs_di_sparse
#endif
#endif
```

immediately before this, which seems to have fixed it, but it's ugly, and I'm not 100% it's probably allowing the structs to be freed properly.  It seems though that there are no tests that end up with these `%extend` functions being called as far I as can tell.  Also, now that there are actually two structs, the calling code would have to be reference the correct one for Siconos.. a bit complicated.  Perhaps we can replace all references to cs_sparse in numerics.py with CSparseMatrix, as defined to the correct type.

**Comment by xhub** [comment]
> However yes I can confirm that changing `max` to `fmax` and not changing `math.h` to `tgmath.h` also works.

Yep sorry, I was concerned about the change from `math.h` to `tgmath.h` that didn't look necessary.

**Comment by xhub** [comment]
> I'll move SuiteSparse configuration stuff to `siconos_find_common_dep`.

Thanks!

**Comment by xhub** [comment]
You're right that I didn't want to generate all the unecessary wrapping code.

I feel like we don't have much choice for wrapping the ``cs_sparse'' struct with SWIG. I don't like that we don't have any control over the struct cs_d?_sparse if it is a system-wide install. But I doubt there would be any change in that one. If we want to be supersafe, we could try to get the different offset of the field in the struct and make sure they match what we expect. I'm not sure that replacing cs_sparse by CSparseMatrix fixes anything, but if it makes things easier, that's good.

Adding a test would be a good thing indeed.

**Comment by xhub** [comment]
> Yes, I proposed this in #188.  I changed the `#include "csparse.h"` in SparseMatrix.h to an opaque `struct cs_sparse`, and the user would have to `#include` it himself to access the struct.  We could still do this with CXSparse perhaps.  Mainly I thought it was a little unusual to include `csparse.h` in the Siconos installed files and for Siconos to export the `cs_` symbols.

Since externals is now a separate library, we have to export it those
symbols in the library. But we shouldn't define those ``by default''.

> That said, I am not sure I understand your concern about using SuiteSparse/CXSparse as a library.  Or should I understand that CXSparse is better to use as a library than CSparse? 

Yes, sorry the formatting wasn't great. But Tim Davis himself says that.

> I don't know much about compiling .mex, does Matlab include its own version of  `cs.h`? I assume by "clash" you mean that both Siconos and Matlab were _exporting_ the cs symbols. If Siconos is now _importing_, I don't see a problem.

The issue is the following: the code in the numerics library needs the cs_* functions. We build those in the externals lib. But MATLAB also exports those symbols. When the cs_* symbols needed by numerics are resolved, those from the MATLAB library are used and ... kaboom.

> By the way speaking of symbol exports, there seems to be a trend in Debian to control symbol exports more carefully, we may have to deal with that.  [(Gcc  "visibility" attribute.)](https://gcc.gnu.org/wiki/Visibility)  It might be a bit of a headache eventually.

That's somethings that I tried to push for: an API.

**Comment by xhub** [comment]
I'm sorry to nitpick a bit, but I would like to have the best possible result here. Two thoughts:
- I would like to replace CSparse by CXSparse in external: we should move to CXSparse everywhere. The only difference would be that we build it if we cannot find one on the system, then we build one.
- Is the SICONOS_INT64 mechanism sensible?

File: numerics/src/FrictionContact/Friction_cst.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (xhub) [reply numerics/src/FrictionContact/Friction_cst.h]: Good thanks

File: numerics/src/LCP/LinearComplementarityProblem_as_ConvexQP.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (xhub) [reply numerics/src/LCP/LinearComplementarityProblem_as_ConvexQP.c]: Looks good

File: numerics/swig/Numerics_typemaps_numericsmatrices.i (hunk 2 above)
[the remark this answers is not in the document]
Comment (xhub) [reply numerics/swig/Numerics_typemaps_numericsmatrices.i]: ok that's sad that SWIG is in the way. For the CS_LONG and SICONOS_INT64 issue, I like the check. I was just worried about the user manually setting CS_LONG, or it being defined somewhere else. With a consistency check, it's all good.

**Comment by radarsat1** [comment]
No apologies for nitpicking, I agree this must be done correctly ;)

> Since externals is now a separate library, we have to export it those symbols in the library. But we shouldn't define those ``by default''.

Okay thanks for the clarification, I did not appreciate this problem.  So the issue is that kernel and numerics uses CSparse and since we build CSparse in a separate library the symbols are needed for kernel/numerics->external link to work.

However, here I am providing a new flag which makes it optionally use an installed CXSparse library.  In the case that USE_SYSTEM_SUITESPARSE=YES, Siconos now does not export any CSparse functions, with the exception of the things in CSparse_extensions.  It does export some functions that take pointers to CSparse structs.  I think that should resolve any problems related to this.

Possibly the CSparse_extensions stuff could be moved to numerics and renamed to SparseMatrix or CSparseMatrix, not sure if it makes a difference, but it would emphasize that these functions belong to Siconos.  Well.. I'll stop talking about that unless I can come up with a decent proposal for it.

Fwiw this patch seems to work for me.  I am only blocked a bit on some memory leak issues, but I'm suspecting these already existed.. so far I can't get the destructor for the Python cs object to be called when I construct the object explicitly, but this is true as well on master actually.

However, I was curious to discover that there were no tests creating a "cs" object, so I wonder if it is actually used/useful?  e.g. functions such as SBM_to_sparse magically return a scipy object (nicely done!). This seems a lot more useful on the Python side.  Conversely, removing that entire `%extend cs_sparse`-related section from numerics_NM.i appears to not break anything at all.  It seems the only way to get a cs() object is to explicitly create it and there are no functions that accept it as an argument.  Is there any motivation to keep it?

e.g. passing a 'cs' object into cs_zentry complains "matrix is not sparse".. indeed, the impl calls cs_convert_from_scipy_sparse immediately on that argument, so it assumes it's a scipy matrix .. the wrapper for struct cs_sparse therefore appears to be obsolete afaict.

> I would like to replace CSparse by CXSparse in external: we should move to CXSparse everywhere. The only difference would be that we build it if we cannot find one on the system, then we build one.

Moving completely to CXSparse should be possible.  I do agree that the built-in CSparse should be used in the case that no system-installed CXSparse is available, that's my goal here.  I think these are two separate things, but you make a good point that we should be sure that SICONOS_USE_CXSPARSE and USE_SYSTEM_SUITESPARSE should not get mixed up, since built-in CSparse may become CXSparse in the future.

Perhaps switching to CXSparse as the built-in library should come before this patch and SICONOS_USE_CXSPARSE should disappear.

> Is the SICONOS_INT64 mechanism sensible?

I think a flag to decide whether Siconos will use cs_dl or cs_di internally does make sense, and this should be in the public config, so yes, I think it's ok.

**Comment by radarsat1** [comment]
I've finally understood something about the problem with the cs() object.  It appears even that %extend constructor gets affected by the typemap on CSparseMatrix.  So the following code actually returns a thin wrapper around a scipy sparse matrix!

```python
def test_SBM_to_sparse2():
    from siconos.numerics import SBM_new_from_file, SBM_to_sparse, cs
    import scipy.sparse
    SBM = SBM_new_from_file(os.path.join(working_dir, 'data/SBM1.dat'))
    r,A = SBM_to_sparse(SBM)
    print('A',A.__class__)
    B = cs(A)
    print('B',B.__class__)
    print('B.this',B.this.__class__)
```
output:
```
('A', <class 'scipy.sparse.csr.csr_matrix'>)
('B', <class 'siconos.numerics.cs'>)
('B.this', <class 'scipy.sparse.coo.coo_matrix'>)
```

So there is really no such representation of `struct cs_sparse` available in Python, it is completely masked by the use of scipy.sparse.  I tried moving `struct cs_sparse` to before `Numerics_typemaps_numericsmatrices.i` is included, but it won't compile.  I propose removing that %extend hack and the `struct cs_sparse`.

**Comment by radarsat1** [comment]
Branch cleaned up, rebased, some new patches.  I propose just adding "Change built-in CSparse to CXSparse" before merging this.  Currently working on it.

**Comment by radarsat1** [comment]
I added a patch that updates to CXSparse from the SuiteSparse 5.0.0 download.  I also added a script that builds cxsparse.c automatically, and found that to make it work with BUILD_AS_CPP, I just had to add `-fpermissive` for that file.  It gives warnings for all the malloc calls, which is annoying, but it means _no changes_ to the upstream sources, which I think is beneficial.  I added a little preamble that was needed to adapt some of the complex numbers code to build properly as C++.

**Comment by radarsat1** [comment]
I'll add, this removes SICONOS_USE_CXSPARSE, since that is now the only way, and USE_SYSTEM_SUITESPARSE still works, so it can be made to use `cs.h` provided on the system.  Currently running tests under various configurations.  I also still need to test this on clang and Mac to make sure it doesn't mess up anyone's work.

**Comment by radarsat1** [comment]
Ah, nope, in my test container I've got some linking errors when using the built-in CXSparse. I'll have to finish this on Monday.

**Comment by radarsat1** [comment]
Seems better now.  Not completely successful with BUILD_AS_CPP.  I could not completely adapt the complex numbers code to C++, so gave up on that for now and skip it in that case, but otherwise it works.  (It is not used by Siconos afaict.)

In any case, non-complex part of CXSparse builds in both cases, and tests are passing for me when BUILD_AS_CPP=NO.  Tested with GCC and Clang with both values for USE_SYSTEM_SUITESPARSE.

**Comment by radarsat1** [comment]
Still occasionally finding corner cases but so far I have compiled and run ctest and examples with most combinations of:

* USE_SYSTEM_SUITESPARSE=YES and NO
* gcc/g++ and clang/clang++
* BUILD_AS_CPP=YES and NO
* mac, ubuntu 14.04, ubuntu 17.04

So I think I will merge this to master at the end of the day and cdash will tell us the rest.

**Comment by radarsat1** [comment]
Any objections to *not installing* these headers?  (Effects of my last patch..)

They seem to me to be "internal" things, not needed by users of Siconos.

numerics:

+  src/tools/open_lib.h
+  src/tools/sanitizer.h
+  src/tools/timers_interf.h
+  src/tools/test_utils.h
+  src/tools/fftw_cycle.h
+  src/tools/tlsdef.h
+  src/tools/numerics_verbose.h
+  src/FrictionContact/Generated/funcodegen.h

kernel:

+ src/utils/SiconosTools/CxxStd.hpp

`numerics_verbose.h` I was not sure, but it requires `tlsdef.h` and it was minimal work to remove it from the Exceptions headers.
